internal/xsocket/xsocket.go

```diff
@@ -0,0 +1,36 @@
+package xsocket
+
+import (
+	"syscall"
+
+	"golang.org/x/sys/unix"
+)
+
+// CloexecSocket creates a new socket with the close-on-exec flag set.
+//
+// If the OS doesn't support the close-on-exec flag, this function will try a workaround.
+func CloexecSocket(domain, typ, proto int) (int, error) {
+	fd, err := unix.Socket(domain, typ|unix.SOCK_CLOEXEC, proto)
+	if err == nil {
+		return fd, nil
+	}
+
+	if err == unix.EINVAL || err == unix.EPROTONOSUPPORT {
```

Comment:
`EINVAL` seems awfully wide an error to catch for this I think, especially with having to take the forklock in the fallback case. 
Can this be further narrowed per-platform with build tags? Afaict linux and all the go-supported bsds support `SOCK_CLOEXEC`, so the only actually relevant platform would be darwin, non-illumos solaris (before 11.4 according to [net source](https://github.com/golang/go/blob/bae2e968e2daadd39c1bdb1221648361d7277ddc/src/net/sock_cloexec_solaris.go#L7) ), maybe AIX?
It looks like `GOOS=darwin` doesn't even define `unix.SOCK_CLOEXEC` so this workaround won't even compile there (would it be necessary? I'm not sure)

I'd be inclined to not support it at all, I don't think any of these platforms are tested at all anyway

Reply:
Darwin should support CLOEXEC I believe starting with some version, but I haven't tested it. 

There's a similar workaround in Go runtime, and this code is loosely based on https://github.com/mdlayher/socket/blob/9c51a391be6c6bc5b7ce52f328497931c5e97733/conn.go#L254-L303

But I'm happy to drop it and simply do `| unix.O_CLOEXEC`

Reply:
yes, `SOCK_CLOEXEC` is missing on Darwin, but other constants are also missing, so not sure how much relevant that is:

```
# github.com/insomniacslk/dhcp/dhcpv4/nclient4
dhcpv4/nclient4/conn_unix.go:42:59: undefined: unix.ETH_P_IP
```

Reply:
@Natolumin do you want me to simplify and just add CLOEXEC directly?